Add masonry-auto-flow to CSS parser.

sammygill opened the pull request:
#### 94e4f6bc3c1bb8c815e01ba47f64923a2f9cc6c2
<pre>
Add masonry-auto-flow to CSS parser.
<a href="https://bugs.webkit.org/show_bug.cgi?id=247078">https://bugs.webkit.org/show_bug.cgi?id=247078</a>
rdar://101597325

Reviewed by Antti Koivisto.

Adds masonry-auto-flow to the CSS parser. masonry-auto-flow has two
values that can both be set. These values can be pack/next
and definite-first/ordered. The initial value is pack if the property
is not specified.

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/masonry-item-placement-001.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/parsing/masonry-parsing-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeMasonryAutoFlow):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::masonryAutoFlow const):
(WebCore::RenderStyle::setMasonryAutoFlow):
(WebCore::RenderStyle::initialMasonryAutoFlow):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/StyleGridData.cpp:
(WebCore::StyleGridData::StyleGridData):
* Source/WebCore/rendering/style/StyleGridData.h:
(WebCore::MasonryAutoFlow::operator== const):
(WebCore::StyleGridData::operator== const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertMasonryAutoFlow):

Canonical link: <a href="https://commits.webkit.org/257566@main">https://commits.webkit.org/257566@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43b81aec5e85a79b7ae67831f0c13e8f39132c7a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99226 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8432 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32353 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108618 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168866 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103225 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8984 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85752 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91725 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106550 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104984 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6784 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90342 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33787 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88604 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21689 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76656 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2312 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23205 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2206 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45606 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5206 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/7455 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42679 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3681 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -3487,6 +3487,7 @@ CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap()
         new DiscretePropertyWrapper<Float>(CSSPropertyFloat, &RenderStyle::floating, &RenderStyle::setFloating),
         new DiscretePropertyWrapper<const Vector<GridTrackSize>&>(CSSPropertyGridAutoColumns, &RenderStyle::gridAutoColumns, &RenderStyle::setGridAutoColumns),
         new DiscretePropertyWrapper<GridAutoFlow>(CSSPropertyGridAutoFlow, &RenderStyle::gridAutoFlow, &RenderStyle::setGridAutoFlow),
+        new DiscretePropertyWrapper<MasonryAutoFlow>(CSSPropertyMasonryAutoFlow, &RenderStyle::masonryAutoFlow, &RenderStyle::setMasonryAutoFlow),
```
🎉

stwrt commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -3258,6 +3258,21 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
     }
+    case CSSPropertyMasonryAutoFlow: {
+        auto list = CSSValueList::createSpaceSeparated();
+        ASSERT((style.isMasonryAutoFlowPack() || style.isMasonryAutoFlowNext()) && (style.isMasonryAutoFlowDefiniteFirst() || style.isMasonryAutoFlowOrdered()));
```
Should this be an || instead of an &&?

[ pack [|](https://drafts.csswg.org/css-values-4/#comb-one) next ] [||](https://drafts.csswg.org/css-values-4/#comb-any) [definite-first | ordered ]

sammygill replied:
Summary from our conversation that I will be making a comment for: 

- RenderStyle holds `m_rareNonInheritedData->grid->masonryAutoFlow`
- `masonryAutoFlow` is 4 bit bitfield 
- This bitfield will be set to one of the enum values of `MasonryAutoFlow`
- Each value in `MasonryAutoFlow` has enough information to know whether it is pack/next *and* definite first/ordered (if ordered is missing when parsing it is implied to act as definite first according to spec behavior)

stwrt commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1226,6 +1227,50 @@ inline GridAutoFlow BuilderConverter::convertGridAutoFlow(BuilderState&, const C
+    return masonryAutoFlow;
+}
+
+
```
extra line not needed

sammygill replied:
Should be fixed now

stwrt commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1226,6 +1227,50 @@ inline GridAutoFlow BuilderConverter::convertGridAutoFlow(BuilderState&, const C
 
+inline MasonryAutoFlow BuilderConverter::convertMasonryAutoFlow(BuilderState&, const CSSValue& value)
+{
+    auto* valueList = dynamicDowncast<CSSValueList>(value);
```
Kind of curious why is this a ptr when the value is a reference?

sammygill replied:
There really wasn't a reason. Changed to use downcast

stwrt reviewed (commented):
Just a few small comments.

anttijk commented on Source/WebCore/rendering/style/StyleGridData.h:
```diff
@@ -133,6 +133,7 @@ class StyleGridData : public RefCounted<StyleGridData> {
     NamedGridLinesMap implicitNamedGridRowLines;
 
     unsigned gridAutoFlow : GridAutoFlowBits;
+    unsigned masonryAutoFlow: GridAutoFlowBits;
```
space before :

Why does this share GridAutoFlowBits with the other field?

sammygill replied:
Changed to use an OpionSet instead

anttijk commented on Source/WebCore/rendering/style/RenderStyleConstants.h:
```diff
@@ -1079,10 +1079,17 @@ static const size_t ColorSchemeBits = 2;
+    InternalMasonryAutoFlowAlgorithmPack = 1 << 0,
+    InternalMasonryAutoFlowAlgorithmNext = 1 << 1,
+    InternalMasonryAutoFlowOrdered       = 1 << 2,
+    InternalMasonryAutoFlowDefiniteFirst = 1 << 3
```
Why is this a single enum that has duplicate values with different names? Shouldn't there be a separate enum for the new things?

Any new enums should be enum classes and it would be nice to convert existing ones too.

sammygill replied:
The two properties were very similar so I tried to share thing where I could (like this and the bits), but I can separate them out

sammygill replied:
Separated into its own enum class now

anttijk commented on Source/WebCore/rendering/style/RenderStyleConstants.h:
```diff
@@ -1092,6 +1099,13 @@ enum GridAutoFlow {
     AutoFlowColumnDense = InternalAutoFlowAlgorithmDense | InternalAutoFlowDirectionColumn
 };
 
+enum MasonryAutoFlow {
```
enum class

sammygill replied:
Changed

anttijk commented on Source/WebCore/rendering/style/RenderStyleConstants.h:
```diff
@@ -1092,6 +1099,13 @@ enum GridAutoFlow {
+    AutoFlowPack = InternalMasonryAutoFlowAlgorithmPack | InternalMasonryAutoFlowDefiniteFirst,
+    AutoFlowPackOrdered = InternalMasonryAutoFlowAlgorithmPack | InternalMasonryAutoFlowOrdered,
+    AutoFlowNext = InternalMasonryAutoFlowAlgorithmNext | InternalMasonryAutoFlowDefiniteFirst,
+    AutoFlowNextOrdered = InternalMasonryAutoFlowAlgorithmNext | InternalMasonryAutoFlowOrdered
```
Please use OptionSets instead of bit math.

sammygill replied:
Changed

anttijk commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -615,6 +616,10 @@ class RenderStyle {
+    bool isMasonryAutoFlowPack() const { return (m_rareNonInheritedData->grid->masonryAutoFlow.contains(MasonryAutoFlow::Pack)); }
+    bool isMasonryAutoFlowNext() const { return (m_rareNonInheritedData->grid->masonryAutoFlow.contains(MasonryAutoFlow::Next)); }
+    bool isMasonryAutoFlowDefiniteFirst() const { return (m_rareNonInheritedData->grid->masonryAutoFlow.contains(MasonryAutoFlow::DefiniteFirst)); }
+    bool isMasonryAutoFlowOrdered() const { return (m_rareNonInheritedData->grid->masonryAutoFlow.contains(MasonryAutoFlow::Ordered)); }
```
Not sure if these individual accessors are necessary. How about just exposing masonryAutoFlow() OptioSet? Doing `masonryAutoFlow().contains(MasonryAutoFlow::Pack)` etc in call sites would not be overly verbose.

sammygill replied:
Removed

mattwoodrow commented on Source/WebCore/rendering/style/StyleGridData.h:
```diff
@@ -133,6 +133,7 @@ class StyleGridData : public RefCounted<StyleGridData> {
     NamedGridLinesMap implicitNamedGridRowLines;
 
     unsigned gridAutoFlow : GridAutoFlowBits;
+    OptionSet<MasonryAutoFlow> masonryAutoFlow;
```
Am I correct that `pack` and `next` are mutually exclusive, and one must be true at all times, and the same is true for `definite-first` and `ordered`?

If so, it seems like this would be better represented as two separate enum values, rather than an OptionSet for both. The OptionSet allows both `pack` and `next` to be expressed at the same time, and I think it'd be clearer if this wasn't possible.

sammygill replied:
Yes, that is how my understanding of it works. The reason I tried to keep it in just `MasonryAutoFlow` was just to have a 1 to 1 correspondence with the spec but we can change it.

How does `masonryAutoFlowPlacementAlgorithm` and `masonryAutoFlowPlacementOrder` sound for the enum values? They are a little long but I would prefer to be a little more verbose if it makes things clearer.

mattwoodrow replied:
They sound great to me!

sammygill replied:
I ended up changing `MasonryAutoFlow` to a struct that contains two enum classes: `MasonryAutoFlowPlacementOrder` and `MasonryAutoFlowPlacementAlgorithm`. Now we just check the value of both of these fields.

mattwoodrow commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -7288,6 +7288,30 @@ RefPtr<CSSValue> consumeGridAutoFlow(CSSParserTokenRange& range)
+    RefPtr<CSSValueList> parsedValues = CSSValueList::createSpaceSeparated();
+    if (packOrNextValue) {
+        CSSValueID packOrNextValueID = packOrNextValue->valueID();
+        if (!definiteFirstOrOrderedValue || (definiteFirstOrOrderedValue && definiteFirstOrOrderedValue->valueID() == CSSValueID::CSSValueDefiniteFirst) || packOrNextValueID == CSSValueID::CSSValueNext)
```
This condition might be easier to read as the inverse set of checks?

If first value is pack, and second value exists and is not definite-first:

mattwoodrow replied:
I'm finding it very weird that the spec has an explicit 'definite-first' value, which differs from the 'grid-auto-flow' value (where 'sparse' is just an implicit default state, but can't ever be specified).

We should never serialise `definite-first` back though, right? We're aiming for shortest possible serialisation and canonical ordering, so it'll always be omitted.

mattwoodrow replied:
Actually, I also find it weird that both grid-auto-flow and masonry-auto-flow are clearly just two different boolean options that are shoved into a single property. If they really insist on doing that, they could at least be consistent with each other though.

sammygill replied:
Yeah it is definitely strange. In terms of changing the conditional, what wouldn't we want to set the value to pack If the second value is `definite-first`? If the first value is `pack` and the second value is `ordered`, then we do not want to append `packOrNextValue`?

mattwoodrow commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1281,6 +1282,53 @@ inline GridAutoFlow BuilderConverter::convertGridAutoFlow(BuilderState&, const C
+        break;
+    case CSSValueDefiniteFirst:
+        masonryAutoFlow.add(MasonryAutoFlow::DefiniteFirst);
+        if (secondValue && secondValue->valueID() == CSSValueNext)
```
This condition isn't possible is it? The only time the parser should have stored an ordered/definite-first value first is if we omitted the implicit 'pack' (which only happens if it was 'Ordered').

Could this be something like:

```
ASSERT(valueList.length() == 1 || 2)
if (valueList.length() == 2) {
    ASSERT(valueList[0].valueID() == Pack || Next);
    ASSERT(valueList[1].validID() == Ordered);
    return { valueList[0].valueID(), Ordered };
} else if (valueList[0].valueID() == CSSValuePack ||  valueList[0].valueID() == CSSValueNext) {
    return { valueList[0].valueID(), DefiniteFirst };
} else {
    ASSERT(valueList[0].validID() == Ordered);
    return { Pack, Ordered };
}
```

I find a lot of the CSS property code complicated since we have a value object that could be an arbitrarily complicated tree of other values, and there's no compile time verification of what 'valid' values are. Trying to structure code that makes it clear what the set of possible values are, with asserts, makes this easier to follow in my opinion :)

sammygill replied:
Agreed I don't think it is possible either. Removed and tried to clean up the logic to make it a little more readable

stwrt commented on Source/WebCore/rendering/style/RenderStyleConstants.h:
```diff
@@ -1089,10 +1089,12 @@ static const size_t ColorSchemeBits = 2;
     InternalAutoFlowDirectionRow    = 1 << 2,
-    InternalAutoFlowDirectionColumn = 1 << 3
+    InternalAutoFlowDirectionColumn = 1 << 3,
+
```
Can we remove the new line here?

sammygill replied:
Should be removed now

sammygill committed:
Add masonry-auto-flow to CSS parser.
https://bugs.webkit.org/show_bug.cgi?id=247078
rdar://101597325

Reviewed by Antti Koivisto.

Adds masonry-auto-flow to the CSS parser. masonry-auto-flow has two
values that can both be set. These values can be pack/next
and definite-first/ordered. The initial value is pack if the property
is not specified.

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/masonry-item-placement-001.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/parsing/masonry-parsing-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeMasonryAutoFlow):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::masonryAutoFlow const):
(WebCore::RenderStyle::setMasonryAutoFlow):
(WebCore::RenderStyle::initialMasonryAutoFlow):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/StyleGridData.cpp:
(WebCore::StyleGridData::StyleGridData):
* Source/WebCore/rendering/style/StyleGridData.h:
(WebCore::MasonryAutoFlow::operator== const):
(WebCore::StyleGridData::operator== const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertMasonryAutoFlow):

Canonical link: https://commits.webkit.org/257566@main
